DOC: update and extend fonts explanation (#23621)
matplotlib/matplotlib: doc/users/explain/fonts.rst

.. redirect-from:: /users/fonts

Fonts in Matplotlib
===================

Matplotlib needs fonts to work with its text engine, some of which are shipped
alongside the installation.  The default font is `DejaVu Sans
<https://dejavu-fonts.github.io>`_ which covers most European writing systems.
However, users can configure the default fonts, and provide their own custom
fonts.  See :doc:`Customizing text properties </tutorials/text/text_props>` for
details and :ref:`font-nonlatin` in particular for glyphs not supported by
DejaVu Sans.

Matplotlib also provides an option to offload text rendering to a TeX engine
(``usetex=True``), see :doc:`Text rendering with LaTeX
</tutorials/text/usetex>`.

> Fonts in PDF and postscript

Reviewer: ```suggestion
Fonts in PDF and PostScript
```

  - Generally contain a much larger character set!
  - Limited Support with Matplotlib

Font Subsetting
~~~~~~~~~~~~~~~

> PDF and postscript support embedded fonts in the output files allowing the

Reviewer: ```suggestion
The PDF and PostScript formats support embedding fonts in files allowing the
```

display program to correctly render the text, independent of what fonts are

> installed on the viewer's computer, without the need to pre-rasterize the text.

Reviewer: ```suggestion
installed on the viewer's computer and without the need to pre-rasterize the text.
```

This ensures that if the output is zoomed or resized the text does not become
pixelated.  However, embedding full fonts in the file can lead to large output
files, particularly with fonts with many glyphs such as those that support CJK
(Chinese/Japanese/Korean).

The solution to this problem is to subset the fonts used in the document and
only embed the glyphs actually used.  This gets both vector text and small
files sizes.  Computing the subset of the font required and writing the new
(reduced) font are both complex problem and thus Matplotlib relies on a
vendored fork of `ttconv <https://github.com/sandflow/ttconv>`_ and `fontTools

> <https://fonttools.readthedocs.io/en/latest/>`__.

Reviewer: I would swap the order of these, so it's clear that `ttconv` is the only vendored fork.

Currently Type 3, Type 42, and TrueType fonts are subseted.  Type 1 fonts are not.

Core Fonts
~~~~~~~~~~

> In addition to the ability to embed fonts, as part of the `postscript

Reviewer: ```suggestion
In addition to the ability to embed fonts, as part of the `PostScript
```

    fig.savefig("AFM_PDF.pdf", format="pdf")
    fig.savefig("AFM_PS.ps", format="ps)

Fonts in SVG
------------

> Text can output to SVG in two ways controlled by the :rc:`svg.fonttype`
> rcparam:

Reviewer: ```suggestion
Text can output to SVG in two ways controlled by :rc:`svg.fonttype`:
```

- as a path (``'path'``) in the SVG
- as string in the SVG with font styling on the element (``'none'``)

When saving via ``'path'`` Matplotlib will compute the path of the glyphs used
as vector paths and write those to the output.  The advantage of this is that
the SVG will look the same on all computers independent of what fonts are
installed.  However the text will not be editable after the fact.
In contrast saving with ``'none'`` will result in smaller files and the
text will appear directly in the markup.  However, the appearance may vary
based on the SVG viewer and what fonts are available.

Fonts in Agg
------------

> To output text to raster formats via Agg Matplotlib relies on `FreeType
> <https://www.freetype.org/>`_.  Because the exactly rendering of the glyphs

Reviewer: ```suggestion
To output text to raster formats via Agg, Matplotlib relies on `FreeType
<https://www.freetype.org/>`_.  Because the exact rendering of the glyphs
```

changes between FreeType versions we pin to a specific version for our image
comparison tests.

How Matplotlib selects fonts
----------------------------

Internally using a Font in Matplotlib is a three step process:

1. a `.FontProperties` object is created (explicitly or implicitly)
2. based on the `.FontProperties` object the methods on `.FontManager` are used

>    to select the closest the "best" font Matplotlib is aware of (except for

Reviewer: ```suggestion
   to select the closest "best" font Matplotlib is aware of (except for
```

   ``'none'`` mode of SVG).
3. the Python proxy for the font object is used by the backend code to render
   the text -- the exact details depend on the backend via `.font_manager.get_font`.

The algorithm to select the "best" font is a modified version of the algorithm
specified by the `CSS1 Specifications
<http://www.w3.org/TR/1998/REC-CSS2-19980512/>`_ which is used by web browsers.
This algorithm takes into account the font family name (e.g. "Arial", "Noto
Sans CJK", "Hack", ...), the size, style, and weight.  In addition to family
names that map directly to fonts there are five "generic font family names" (
serif, monospace, fantasy, cursive, and sans-serif) that will internally be
mapped to any one of a set of fonts.

Currently the public API for doing step 2 is `.FontManager.findfont` (and that
method on the global `.FontManager` instance is aliased at the module level as

> `.font_manager.findfont`) will only find a single font and return the absolute

Reviewer: ```suggestion
`.font_manager.findfont`), which will only find a single font and return the absolute
```

path to the font on the filesystem.

Font Fallback
-------------

> There is no font that covers the unicode space thus it is possible for the

Reviewer: ```suggestion
There is no font that covers the entire Unicode space thus it is possible for the
```

users to require a mix of glyphs that can not be satisfied from a single font.
While it has been possible to use multiple fonts within a Figure, on distinct
`.Text` instances, it was not previous possible to use multiple fonts in the
same `.Text` instance (as a web browser does).  As of Matplotlib 3.6 the Agg,
SVG, PDF, and PS backends will "fallback" through multiple fonts in a single
`.Text` instance:

.. plot::
   :include-source:
   :caption: The string "There are 几个汉字 in between!" rendered with 2 fonts.

   fig, ax = plt.subplots()
   ax.text(
       .5, .5, "There are 几个汉字 in between!",
       family=['DejaVu Sans', 'WenQuanYi Zen Hei'],
       ha='center'
   )

Internally this is implemented by setting The "font family" on

> `.FontProperties` objects to a list of font families.  Using a (currently)
> private API extract a list of paths to all of the fonts found and then
> construct a single `.ft2font.FT2Font` object that is aware of all of the fonts.

Reviewer: ```suggestion
`.FontProperties` objects to a list of font families.  A (currently)
private API extracts a list of paths to all of the fonts found and then
constructs a single `.ft2font.FT2Font` object that is aware of all of the fonts.
```
